// @mysten/seal/dist/cjs/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var index_exports = {};
__export(index_exports, {
  AesGcm256: () => import_aes.AesGcm256,
  EncryptedObject: () => import_types.EncryptedObject,
  KeyStore: () => import_key_store.KeyStore,
  SessionKey: () => import_session_key.SessionKey,
  encrypt: () => import_encrypt.encrypt,
  getAllowlistedKeyServers: () => import_key_server.getAllowlistedKeyServers,
  retrieveKeyServers: () => import_key_server.retrieveKeyServers,
  verifyKeyServer: () => import_key_server.verifyKeyServer
});
module.exports = __toCommonJS(index_exports);
var import_aes = require("./aes.js");
var import_encrypt = require("./encrypt.js");
var import_key_server = require("./key-server.js");
var import_key_store = require("./key-store.js");
var import_session_key = require("./session-key.js");
var import_types = require("./types.js");
