// @mysten/seal/dist/cjs/types.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var types_exports = {};
__export(types_exports, {
  Ciphertext: () => Ciphertext,
  EncryptedObject: () => EncryptedObject,
  IBEEncryptions: () => IBEEncryptions
});
module.exports = __toCommonJS(types_exports);
var import_bcs = require("@mysten/sui/bcs");
const IBEEncryptions = import_bcs.bcs.enum("IBEEncryptions", {
  BonehFranklinBLS12381: import_bcs.bcs.struct("BonehFranklinBLS12381", {
    encapsulation: import_bcs.bcs.bytes(96),
    shares: import_bcs.bcs.vector(import_bcs.bcs.bytes(32))
  })
});
const Ciphertext = import_bcs.bcs.enum("Ciphertext", {
  Aes256Gcm: import_bcs.bcs.struct("Aes256Gcm", {
    blob: import_bcs.bcs.vector(import_bcs.bcs.U8),
    aad: import_bcs.bcs.option(import_bcs.bcs.vector(import_bcs.bcs.U8))
  }),
  Plain: import_bcs.bcs.struct("Plain", {})
});
const EncryptedObject = import_bcs.bcs.struct("EncryptedObject", {
  version: import_bcs.bcs.U8,
  package_id: import_bcs.bcs.bytes(32),
  id: import_bcs.bcs.vector(import_bcs.bcs.U8),
  services: import_bcs.bcs.vector(import_bcs.bcs.tuple([import_bcs.bcs.bytes(32), import_bcs.bcs.U8])),
  threshold: import_bcs.bcs.U8,
  encrypted_shares: IBEEncryptions,
  ciphertext: Ciphertext
});
